### DISCUSSION TITLE
Working where I need to merge something to test

### DISCUSSION THREAD

**Discussion by d3zd3z** [jj-vcs/jj#3490] [Q&A]
I'm working on a couple of changes that are being pushed to github. One of these tests fixes a problem with my test framework, and I need this change to be able to work on the other changes. But, to github, I want them to remain independent.

What I'm doing now is that I basically work, until I think I'm ready to test. Then I do the following:
```bash
jj new @ test-fix
cargo test ...
jj abandon @
```

This seems to work, as long as I'm careful.

I've also tried just making new work on top of the merge, and then rebasing the work "around" the merge back to the main tree. But, I found it harder to get rid of the merge then.

Any suggestions on how to do this better?

**Comment by ilyagr** [comment]
I'd probably consider the same two options. 

> ... then rebasing the work "around" the merge back to the main tree. But, I found it harder to get rid of the merge then.

What do you mean? If I wanted to get rid of the merge in this situation, I'd first get rid of all of its children.

The `jj abandon` behavior recently changed WRT merges, in the direction of making it harder to get rid of a merge. I think we could consider adding some option to `jj abandon` than nukes merges, but I haven't had enough experience myself with these kinds of problems to have a clear enough understanding or opinion.

**Reply by d3zd3z** [reply]
The reason I want to get rid of the merge is to effectively discard all of its children.

another workflow that would work, but I couldn’t figure out how to do would be something that effectively squashes the entire merge into a single commit. This would be easier to move around and discard once the upstream merges that work. But I haven’t figured out how to squash all of the changes, basically, I want to squash main..otherbranch, and apply it to my current work.

**Reply by martinvonz** [reply]
> The reason I want to get rid of the merge is to effectively discard all of its children.

`jj abandon xyz::` will abandon a commit and its descendants.

> another workflow that would work, but I couldn’t figure out how to do would be something that effectively squashes the entire merge into a single commit.

`jj squash --from` accepts multiple commits. Maybe `jj squash --from main..otherbranch` is what you're looking for (it squashes it into the working copy by default).

**Reply by ilyagr** [reply]
What I said before may have been a bit misleading. On second thought, `jj abandon` should work fine and "nuke merges".

I thought that the merge might be preserved if the merge commit was also the working copy, but it seems to nuke the merge even in this case upon experimentation.

So, try the things Martin suggested and let us know if something still doesn't work.

If I confused you before, sorry about that. If I'm confusing you now, you can try ignoring both of my messages and seeing if that helps :)

**Reply by martinvonz** [reply]
> I thought that the merge might be preserved if the merge commit was also the working copy, but it seems to nuke the merge even in this case upon experimentation.

I actually plan to make it that way. I think it's more consistent to not treat merge commits differently. It also means you can `jj abandon` to restart your current work even if your working copy is a merge commit.

**Reply by ilyagr** [reply]
Then we should change the definition of a "discardable" commit so that `jj abandon && jj new somewhere_else` does kill the merge.
